### PR TITLE
Implementation resuscitating cells

### PR DESCRIPTION
**Pull request by Gabbendorf**
Implemented method evolve so that for the next generation it adds to the list of cells both survived and resuscitated cells. Dead cells (empty positions) resuscitate if they have exactly 3 live cells as neighbours. Live cells survive if they have either 2 or 3 live cells as neighbours, otherwise they die.
Please feel free to make any comments. I've found very hard to implement this last feature (cells resuscitating), could this be because of a bad-thought design? Also refactoring was very complicated. Any suggestions?
I would like to have feedback re. naming methods/variable, commits, besides any other comments  of course, as well. Are these names too long/complicated? Is it ok to commit after just changing a method name? Etc.
Should I have used private methods as well? 
Thanks for any comments!

### CODE REVIEW

File: lib/world.rb [hunk 1]
```diff
@@ -8,13 +8,28 @@ def initialize(cells)
...
-      }
-      @cells = next_generation
+    @cells.each do |cell|
+      if survive?(cell.position)
+        next_generation.push(cell)
+      end
+    end
+    positions_of_potential_resuscitating_cells.each do |position|
+      if resuscitate?(position)
+        new_cell = new_instance_of_cell(position)
```
Comment (RabeaGleissner) [inline lib/world.rb]: I think it would be better not to wrap `Cell.new(position) into a method. It's a bit unnecessary and just makes the code a bit more difficult to read. It's obvious was `Cell.new` does, so no need to give it a name by wrapping it in a method.

File: lib/world.rb [hunk 2]
```diff
@@ -37,8 +52,28 @@ def count_live_neighbours(position)
...
+
+  def cells_positions
+    @cells.map {|cell| cell.position }
+  end
+
+  def number_of_neighbours_to_survive(count)
     count == 2 || count == 3
   end
 
+  def number_of_neighbours_to_resuscitate(count)
```
Comment (RabeaGleissner) [inline lib/world.rb]: This method returns a boolean so the method name should end with a `?`

Comment (RabeaGleissner) [reply lib/world.rb]: I think the name isn't quite right. It sounds like it returns the number of neighbours necessary to resuscitate. But instead it checks if the neighbour count is correct to resuscitate the cell. Maybe change the name to `enough_neighbours_to_resuscitate?`

**Review by RabeaGleissner — COMMENTED** [review]
I haven't quite managed to go through all of it. Will do more later but I thought I'd already submit my comments now.

File: lib/world.rb (hunk 1 above)
Comment (Gabbendorf) [reply lib/world.rb]: Gotcha

File: lib/world.rb (hunk 2 above)
Comment (Gabbendorf) [reply lib/world.rb]: Previous name was ideal_neighbours_to_survive, something like that, I didn't put the question mark though. Would that have been better?

File: spec/world_spec.rb [hunk 3]
```diff
@@ -5,58 +5,100 @@
 
   let (:world) {World.new}
 
-  it "returns no live cells if there is only 1 cell" do
-    cell = Cell.new([1,1])
-    world = World.new([cell])
+  describe "applies the first 3 rules concerning live cells" do
```
Comment (RabeaGleissner) [inline spec/world_spec.rb:8]: I think the describe blocks aren't very symmetrical here. Either use a describe block for each rule or none at all. But clustering three together and then having one alone doesn't make much sense on first glance. Maybe you did it because for the first three rules have to do with alive cells and the last rule with a dead cell? Then you should call the describe block message something like `describe 'alive cell survives' do` and `describe 'dead cell comes alive' do` or something along those lines.

File: spec/world_spec.rb [hunk 4]
```diff
@@ -5,58 +5,100 @@
...
+    it "returns no live cells if there is only 1 cell" do
+      cell = Cell.new([1,1])
+      world = World.new([cell])
 
-    world.evolve
+      world.evolve
 
-    expect(world.cells).to eq([])
-  end
+      expect(world.cells).to eq([])
```
Comment (RabeaGleissner) [inline spec/world_spec.rb]: As you will probably read in the '4 simple rules of design' book, it would be nice to have a method `empty?` on the World class. Then instead of checking if cells is an empty array, you could say `expect(world.empty?).to be true`

File: spec/world_spec.rb [hunk 5]
```diff
@@ -5,58 +5,100 @@
...
+    end
   end
 
-  it "returns how many live cells a position has around" do
-    first_cell = Cell.new([0,0])
-    second_cell = Cell.new([0,1])
-    third_cell = Cell.new([1,0])
-    world = World.new([first_cell, second_cell, third_cell])
+    it "returns how many live cells a position has around" do
```
Comment (RabeaGleissner) [inline spec/world_spec.rb]: The wording could be improved here. For example: `it "returns the number of live neighbours" do`

File: spec/world_spec.rb [hunk 6]
```diff
@@ -5,58 +5,100 @@
...
+
+  describe "applies the last rule concerning dead cells" do
+    it "resuscitates dead cells if they have exactly 3 live neighbours" do
+      first_cell = Cell.new([0,0])
+      second_cell = Cell.new([1,0])
+      third_cell = Cell.new([0,1])
+      fourth_cell = Cell.new([2,-1])
+      fifth_cell = Cell.new([-1,-1])
+      world = World.new([first_cell, second_cell, third_cell, fourth_cell, fifth_cell])
+      expect(world.count_live_neighbours([1,1])).to eq(3)
```
Comment (RabeaGleissner) [inline spec/world_spec.rb]: I think this test is a bit too big. As with methods, it is better to have smaller tests. 
This test could be split into two tests:
- one test to assert how many neighbours a cell has
- one test to assert that when you start with three cells that are neighbouring  one dead cell, you end up with four cells

File: spec/world_spec.rb [hunk 7]
```diff
@@ -5,58 +5,100 @@
...
+      expect(world.count_live_neighbours([-1,0])).to eq(3)
+      expect(world.count_live_neighbours([0,-1])).to eq(3)
+      expect(world.count_live_neighbours([1,-1])).to eq(3)
+
+      world.evolve
+
+      expect(world.cells_positions).to contain_exactly([0,0], [1,0], [0,1], [1,1], [1,-1],[0,-1],[-1,0])
+    end
+
+    it "returns positions around live cells where new cells could resuscitate" do
```
Comment (RabeaGleissner) [inline spec/world_spec.rb]: yep, this is a good one ✅

File: spec/world_spec.rb [hunk 8]
```diff
@@ -5,58 +5,100 @@
...
+      expect(potential_resuscitating_cells).to contain_exactly([-1,1],[-1,2],[0,2],[1,2],[1,0],[1,-1],[0,-1],[-1,-1],[-1,0],[2,2],[2,1],[2,0])
+    end
+  end
 
-    neighbours_first_cell = world.count_live_neighbours([0,0])
-    neighbours_second_cell = world.count_live_neighbours([0,1])
-    neighbours_third_cell = world.count_live_neighbours([1,0])
-    neighbours_empty_position = world.count_live_neighbours([1,2])
-    neighbours_empty_position2 = world.count_live_neighbours([2,2])
+    it "returns positions of all live cells" do
```
Comment (RabeaGleissner) [inline spec/world_spec.rb]: This one is good too. But I would probably put it somewhere higher up. It seems like this test would come first and once you have the method `cells_positions` nailed down, then you can build on it and get the other methods working.

File: lib/world.rb [hunk 9]
```diff
@@ -8,13 +8,28 @@ def initialize(cells)
...
+      if resuscitate?(position)
+        new_cell = new_instance_of_cell(position)
+        next_generation.push(new_cell)
+      end
+    end
+    @cells = next_generation
+  end
+
+  def survive?(cell_position)
+    neighbours_count = count_live_neighbours(cell_position)
```
Comment (RabeaGleissner) [inline lib/world.rb]: The method name `count_live_neighbours` can be shortened to `neighbour_count`. You're only counting the live cells as neighbours anyway, aren't you?

File: lib/world.rb (hunk 2 above)
Comment (RabeaGleissner) [reply lib/world.rb]: Yes, that would have been a bit better. Basically, a method that returns a boolean, should have a name that can be answered with 'true' or 'false'.

**Review by RabeaGleissner — COMMENTED** [review]
Hey Gabi! I added a few more comments. Overall, really good! I like that you split out a lot of methods. Sometimes the method names are a bit too wordy. But I guess with more practise you might be able to find shorter names. 
Enjoy your weekend 🌞

File: spec/world_spec.rb (hunk 3 above)
Comment (Gabbendorf) [reply spec/world_spec.rb:8]: I've mentioned in describe that the first describe block  is 'concerning alive cells' and the last 'concerning dead cells': is this not good enough? Yes, it looked weird to me too having a describe block with less tests. I kept them regardless as I was curious on your comments about that :)

File: spec/world_spec.rb (hunk 4 above)
Comment (Gabbendorf) [reply spec/world_spec.rb]: Nice!
